### ISSUE TITLE
[Feat]: Configure hooks directly in workflow step (i.e. without `templatesync.yml`)

### ISSUE THREAD

**Issue by kdeldycke** [AndreasAugustin/actions-template-sync#467]
### Describe the feature

`actions-template-sync` supports hooks and that's great! I'm planning to use them to run a couple of `sed` commands in the `precommit` phase, so I can adapt to the target repository some hard-coded strings from the source template repository.

For this I need a `templatesync.yml`. What I'd love to have is a way to encode these rules directly on the workflow step.

I.e. I'd like to write this:
```yaml
on:
  (...)

jobs:
  repo-sync:
    (...)

    steps:
      - uses: actions/checkout@v4
        
      - name: actions-template-sync
        uses: AndreasAugustin/actions-template-sync@v1.6.2
        with:
          github_token: ${{ secrets.GITHUB_TOKEN }}
          source_repo_path: <owner/repo>
          upstream_branch: <target_branch> 
          pr_labels: <label1>,<label2>[,...] 
          is_allow_hooks: true
          hooks:
            precommit:
              commands:
                # Replace "/kdeldycke/awesome-template/" in URLs by "/kdeldycke/awesome-iam/"
                - sed -i "s/\/kdeldycke\/awesome-template\//\/kdeldycke\/awesome-iam\//g" ./*.md
```

And bypass entirely the creation of a `templatesync.yml` file.

### Use Case

I have a [collection of `kdeldycke/awesome-*` repositories](https://github.com/kdeldycke?tab=repositories&q=awesome-), in which I push content from my [`kdeldycke/awesome-template` repository](https://github.com/kdeldycke/awesome-template).

To do that I [rely on a reused workflow](https://github.com/kdeldycke/workflows/blob/f6397296baae2711862c05ef65e46470fe7921f1/.github/workflows/docs.yaml#L373-L389), that is [called in the target `awesome-*` repositories](https://github.com/kdeldycke/awesome-iam/blob/b84f9e1a58c5f8c5c0cb0af4197c61c9704ec6c5/.github/workflows/docs.yaml#L12-L13).

With the current version of `actions-template-sync`, I have to add an maintain a dedicated `templatesync.yml` file in each target repository.

But with the addition of hook rules directly in the workflow step, I can create generic `sed` commands in my reuseable workflows. That's extremely powerful to automate tedious work.

### Proposed Solution

Add support for the `hooks` > `precommit` > `commands` keywords under the `with` parameter.

### Notes

This is a follow up on:
- #440
- #439

### Acknowledgements

- [ ] I may be able to implement this feature request

**Comment by AndreasAugustin** [comment]
Hi @kdeldycke and thanks for the feature request.
If I remember right the reason for having the file approach was the missing possibility to use objects within GitHub action input parameters.

https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#inputs

If there haven't been changes lately the parameters within the `with` need to be a string.

Maybe 🤔 it is possible to split the current action so that you are able to execute the hooks even in another action. Not super sure if this is working (need to check how the filesystem is acting within the dockerized action)

I definitely need to do some research first.

**Comment by kdeldycke** [comment]
> Maybe 🤔 it is possible to split the current action so that you are able to execute the hooks even in another action. Not super sure if this is working (need to check how the filesystem is acting within the dockerized action)

Yeah, that might be a workaround I'm ready to explore. Like having a second step right after `actions-template-sync` to amend the PR it created. But for this to work I would need at least the `actions-template-sync` action to output the branch name and/or PR number.

Here is how it would look like:
```yaml
    steps:
      - uses: actions/checkout@v4.1.1

      - name: Sync from template repo
        id: template_sync
        uses: AndreasAugustin/actions-template-sync@v1.6.2
        with:
          # We need custom PAT with workflows permissions so we can update all the boilerplate .github files from
          # awesome-template.
          github_token: ${{ secrets.WORKFLOW_UPDATE_GITHUB_PAT || secrets.GITHUB_TOKEN }}
          source_repo_path: kdeldycke/awesome-template
          pr_title: "[sync] Update from `awesome-template` repository"
          pr_commit_msg: "[sync] Update from `awesome-template` repository"
          pr_labels: "📚 documentation"

      - name: Checkout template sync PR
        uses: actions/checkout@v4.1.1
        with:
          ref: ${{ steps.template_sync.outputs.pr_branch }}

      - name: Update repo URLs
        # Replace "/kdeldycke/awesome-template/" in URLs by "/kdeldycke/awesome-<repo_id>/".
        run: >
          find ./.github/ -iname "*.md" -or -iname "*.yaml" -exec
          sed -i "s/\/kdeldycke\/awesome-template\//\/kdeldycke\/${{ github.event.repository.name }}\//g" "{}" \;

      - name: Push to PR
        uses: peter-evans/create-pull-request@v6.0.0
        with:
          assignees: ${{ github.actor }}
          commit-message: "[sync] Update template URLs"
          branch: ${{ steps.template_sync.outputs.pr_branch }}

```

So as you can see, I would need a `steps.template_sync.outputs.pr_branch` for this to work in theory.

**Comment by AndreasAugustin** [comment]
Hi @kdeldycke .
As a current workaround you can do the following starting with version `v1.7.0` (#469)

```yaml
# using GITHUB_TOKEN here, please adjust to your needs
- name: Sync from template repo
        id: template_sync
        uses: AndreasAugustin/actions-template-sync@v1.7.0
       env:
         MY_ENV_VARIABLE_NAME: awesome-iam  # is possible to be used within the templatesync.yml
        with:
          github_token: ${{ secrets.GITHUB_TOKEN }}
          source_repo_path: kdeldycke/awesome-template
          pr_title: "[sync] Update from `awesome-template` repository"
          pr_commit_msg: "[sync] Update from `awesome-template` repository"
          pr_labels: "📚 documentation"
          is_allow_hooks: true
```

(still need the file)
and the file `templatesync.yml`

```yaml
 hooks:
            precommit:
              commands:
                # Replace "/kdeldycke/awesome-template/" in URLs by "/kdeldycke/awesome-iam/"
                - sed -i "s/\/kdeldycke\/awesome-template\//\/kdeldycke\/${MY_ENV_VARIABLE_NAME}\//g" ./*.md
```

I think this could solve your current use case. Nevertheless I will think about the other options.

**Comment by AndreasAugustin** [comment]
@all-contributors please add @kdeldycke for idea

**Comment by kdeldycke** [comment]
I experimented yesterday with the new `pr_branch` but stumbled upon an issue of it being empty: https://github.com/AndreasAugustin/actions-template-sync/issues/476

**Comment by AndreasAugustin** [comment]
> I experimented yesterday with the new `pr_branch` but stumbled upon an issue of it being empty: #476

yes, sorry. Did not consider those 'edge-cases'. #479 (`v1.8.1`) should fix that issue.
Please consider further edge cases (https://github.com/AndreasAugustin/actions-template-sync#action-outputs), but those should not be sufficient for your current issue (related to the action run log).

There is likely to be a further feature (soon) #478 which will likely give some value to your workaround solution (actually that was the feature I meant with the `split` I mentioned within my comment above).

Just out of curiosity.
Have you also considered using the `env` variable solution I mentioned above? The idea is to use something like
(Remark within https://github.com/AndreasAugustin/actions-template-sync#lifecycle-hooks) mentioned here: https://github.com/AndreasAugustin/actions-template-sync/issues/467#issuecomment-1970691769

**Comment by kdeldycke** [comment]
Thanks for identifying and listing the edge-case regarding `pr_branch`! To be honest these edge-cases are surprising for newcomers and I predict we'll get some tickets from users about these. 😅 But I'm happy with the current state of things and I'll find a way to workaround current limitations.

And good ideas about #478. That's a powerful feature! 💪

As for the `env` way, I'm not too fond of that. The syntax is awkward as it's bash syntax (`${MY_VAR}`) embedded in a YAML, where I expect to find GitHub's own `${{ }}` string interpolation grammar. It too much cognitive overload to my small brain ahah! 😁

This, and I don't want to copy-n-past, keep in sync and maintain another Nth CI/CD file at the root of my repositories. My policy is: move every repository-specific config in a `pyproject.toml` or hide the gory details in a reused workflow. So if I go with `templatesync.yml`[^1], I'll have to generate it dynamically in-place, the way I [generate a `ruff.toml` file](https://github.com/kdeldycke/workflows/blob/646264a4be69821616a9850860e7d5ab5a037e01/.github/workflows/autofix.yaml#L168-L217) then [remove it](https://github.com/kdeldycke/workflows/blob/646264a4be69821616a9850860e7d5ab5a037e01/.github/workflows/autofix.yaml#L229-L231) before [committing](https://github.com/kdeldycke/workflows/blob/646264a4be69821616a9850860e7d5ab5a037e01/.github/workflows/autofix.yaml#L247-L263).[^2]


[^1] BTW, speaking of edge-cases, do you recognize the `.yaml` extension in addition to `.yml`? 😁

[^2] Another alternative would be to support remote `templatesync.yml`, like so:
```yaml
```

**Comment by kdeldycke** [comment]
It took me [a week and 45 commits](https://github.com/kdeldycke/awesome-billing/compare/4add3b1...e0aa75a), but I finally found a workaround to do what I was looking for.

Tl;Dr:
```yaml
  awesome-template-sync:
    name: Sync awesome template
    if: >
      startsWith(github.event.repository.name, 'awesome-')
      && github.event.repository.name != 'awesome-template'
    runs-on: ubuntu-22.04
    # We need custom PAT through the whole job so we get workflow permissions to update all the boilerplate .github
    # files from awesome-template.
    steps:

      - name: Initial checkout
        uses: actions/checkout@v4.1.1
        with:
          token: ${{ secrets.WORKFLOW_UPDATE_GITHUB_PAT || secrets.GITHUB_TOKEN }}
          fetch-depth: 0

      - name: Sync from template repo
        id: template_sync
        uses: AndreasAugustin/actions-template-sync@v1.8.1
        with:
          github_token: ${{ secrets.WORKFLOW_UPDATE_GITHUB_PAT || secrets.GITHUB_TOKEN }}
          source_repo_path: kdeldycke/awesome-template
          pr_title: "[sync] Updates from `awesome-template`"
          pr_commit_msg: "[sync] Updates from awesome-template"
          pr_branch_name_prefix: "sync-awesome-template"
          pr_labels: "📚 documentation"

      - name: Checkout new template sync branch
        uses: actions/checkout@v4.1.1
        with:
          token: ${{ secrets.WORKFLOW_UPDATE_GITHUB_PAT || secrets.GITHUB_TOKEN }}
          ref: ${{ steps.template_sync.outputs.pr_branch }}
          fetch-depth: 0

      # XXX We cannot rely on peter-evans/create-pull-request to manage the push of the new commit to the PR, as it
      # ends up with a: "Unexpected error: fatal: could not open '.git/COMMIT_EDITMSG': Permission denied".
      # See: https://github.com/AndreasAugustin/actions-template-sync/issues/484
      # So we manage the next steps with bare git calls.

      - name: Update repo URLs
        # Replace "/kdeldycke/awesome-template/" in URLs by "/kdeldycke/awesome-<repo_id>/".
        run: >
          find ./.github/ -type f -iregex ".*\.\(md\|yaml\)$" -print -exec sed -i
          "s/\/kdeldycke\/awesome-template\//\/kdeldycke\/${{ github.event.repository.name }}\//g" "{}" \;

      - name: Deactivate LFS
        # XXX Was introduced in https://github.com/AndreasAugustin/actions-template-sync/commit/5335a55
        run: |
          git config lfs.https://github.com/kdeldycke/${{ github.event.repository.name }}.git/info/lfs.locksverify false

      - name: Setup Git user
        run: |
          git config --global user.name "${{ github.actor }}"
          git config --global user.email "${{ github.actor }}@users.noreply.github.com"

      - name: Merge change to previous commit
        run: |
          git commit --all --amend --no-edit

      - name: Push new commit to PR
        run: |
          git push --force
```

This job is available at: https://github.com/kdeldycke/workflows/blob/ac65c9b5a5878d9b7c301cb5166302c147be55e3/.github/workflows/docs.yaml#L367-L419

**Comment by kdeldycke** [comment]
Note that my workaround above is only working after a second pass. The fist pass always ends up on this error:
```yaml
  git commit --all --amend --no-edit
  shell: /usr/bin/bash -e {0}
error: insufficient permission for adding an object to repository database .git/objects
error: .github/ISSUE_TEMPLATE/new-link.yaml: failed to insert into database
error: unable to index file '.github/ISSUE_TEMPLATE/new-link.yaml'
fatal: updating files failed
Error: Process completed with exit code 128.
```

![Screenshot 2024-03-04 at 15 13 31](https://github.com/AndreasAugustin/actions-template-sync/assets/159718/e58e797b-45ba-4d00-b6ca-0577a0434f23)

This might be related to: https://github.com/AndreasAugustin/actions-template-sync/issues/484

The full log of that workflow run is available at: https://github.com/kdeldycke/awesome-billing/actions/runs/8139058758/job/22241336465

**Comment by AndreasAugustin** [comment]
@kdeldycke thanks for the input and super sorry for the hassle you are having getting this to work.
#489 should be the part you kinda looking for.
Sad part: Was not able to define the hooks as `yaml` object (seems not supported by github actions so far). You need to use a `yaml` string (have a prefix `>` after the `hooks` input parameter), e.g.

```yaml
- name: Test action step
  uses: AndreasAugustin/actions-template-sync@v1
  env:
    MY_VAR: "foo"  # possible to define envrionment variables
  with:
    source_repo_path: AndreasAugustin/template.git
    upstream_branch: main
    is_dry_run: true
    is_allow_hooks: true
    hooks: >
      prepull:
        commands:
          - echo 'hi, we are within the prepull phase'
          - echo 'maybe you want to do adjustments on the local code'
```

Hope this is also ok for your use case. (unfortunately I did not think about this solution at first thoughts... )

**Comment by AndreasAugustin** [comment]
> Thanks for identifying and listing the edge-case regarding `pr_branch`! To be honest these edge-cases are surprising for newcomers and I predict we'll get some tickets from users about these. 😅 But I'm happy with the current state of things and I'll find a way to workaround current limitations.
> 

Yes, kinda agree :+1: 
The thing is that we have the following possible flows and therefore those edge cases :smile_cat: (open for discussions)

- template changes already merged -> in this case it does not make sense to open a PR because those changes are already part of the history
- template changes are already in a PR -> in this case the action finishes because the PR is already open. (Your use case #482 not considered yet).

> And good ideas about #478. That's a powerful feature! 💪
> 

Thanks :blush: will try to implement this soon. Thanks btw for your found bugs. Think I need to do some investigations before fully implement the feature.

> As for the `env` way, I'm not too fond of that. The syntax is awkward as it's bash syntax (`${MY_VAR}`) embedded in a YAML, where I expect to find GitHub's own `${{ }}` string interpolation grammar. It too much cognitive overload to my small brain ahah! 😁
> 

understood, but as a workaround would have possibly solve the issue faster :ninja:  (but am super happy about the work and the mentioned bugs) . Agree on the bash interpolation part (YAML has a small difference within variable interpolation syntax)

> This, and I don't want to copy-n-past, keep in sync and maintain another Nth CI/CD file at the root of my repositories. My policy is: move every repository-specific config in a `pyproject.toml` or hide the gory details in a reused workflow. So if I go with `templatesync.yml`[^1], I'll have to generate it dynamically in-place, the way I [generate a `ruff.toml` file](https://github.com/kdeldycke/workflows/blob/646264a4be69821616a9850860e7d5ab5a037e01/.github/workflows/autofix.yaml#L168-L217) then [remove it](https://github.com/kdeldycke/workflows/blob/646264a4be69821616a9850860e7d5ab5a037e01/.github/workflows/autofix.yaml#L229-L231) before [committing](https://github.com/kdeldycke/workflows/blob/646264a4be69821616a9850860e7d5ab5a037e01/.github/workflows/autofix.yaml#L247-L263).[^2]
> 
> [^1] BTW, speaking of edge-cases, do you recognize the `.yaml` extension in addition to `.yml`? 😁
> 

you got me :ninja: 

> [^2] Another alternative would be to support remote `templatesync.yml`, like so:

Yes, nice idea. Maybe also (or in addition) git submodules is possible.

> 
> ```yaml
>       - name: actions-template-sync
>         uses: AndreasAugustin/actions-template-sync@v1
>         with:
>           (...)
>           templatesync: https://raw.githubusercontent.com/kdeldycke/workflows/v3.0.0/templatesync.yml
> ```

**Comment by kdeldycke** [comment]
While the hooks are not proper YAML objects, having a way to add them directly in the action is a great progress! Thanks to this new parameter, I simplified a lot my workflow and everything is now self-contained within the `actions-template-sync` step.

Thanks a lot @AndreasAugustin for listening and considering my use-cases. `actions-template-sync` is a fantastic action and my little investment in it will pay off in the future by reducing the maintenance burden on my project. 💪
